File: apps/api/src/domain/services/d-day-calculator.service.ts
import { Injectable } from '@nestjs/common';

@Injectable()
export class DDayCalculatorService {
  calculateDaysFromStart(
    startDate: Date,
    referenceDate: Date = new Date(),
  ): number {
    const diffTime = referenceDate.getTime() - startDate.getTime();
    return Math.floor(diffTime / (1000 * 60 * 60 * 24));
  }

  calculateDaysUntil(
    targetDate: Date,
    referenceDate: Date = new Date(),
  ): number {
    const diffTime = targetDate.getTime() - referenceDate.getTime();
    return Math.ceil(diffTime / (1000 * 60 * 60 * 24));
  }

  getNextAnniversary(startDate: Date, referenceDate: Date = new Date()): Date {
    const anniversary = new Date(startDate);
    anniversary.setFullYear(referenceDate.getFullYear());

    if (anniversary < referenceDate) {
      anniversary.setFullYear(referenceDate.getFullYear() + 1);
    }

    return anniversary;
  }

  getAnniversaryNumber(
    startDate: Date,
    referenceDate: Date = new Date(),
  ): number {
    return referenceDate.getFullYear() - startDate.getFullYear();
  }

  getUpcomingMilestones(
    startDate: Date,
    referenceDate: Date = new Date(),
    count: number = 3,
  ): { days: number; date: Date }[] {
    const daysFromStart = this.calculateDaysFromStart(startDate, referenceDate);
    const milestones: { days: number; date: Date }[] = [];

    const intervals = [100, 200, 300, 365, 500, 730, 1000, 1095, 1461];

    for (const interval of intervals) {
      if (interval > daysFromStart && milestones.length < count) {
        const milestoneDate = new Date(startDate);
        milestoneDate.setDate(milestoneDate.getDate() + interval);
        milestones.push({ days: interval, date: milestoneDate });
      }
    }

    return milestones;
  }
}

File: apps/api-kotlin/src/main/kotlin/com/couplecalendar/domain/service/DDayCalculatorService.kt
package com.couplecalendar.domain.service

import org.springframework.stereotype.Service
import java.time.LocalDate
import java.time.temporal.ChronoUnit

@Service
class DDayCalculatorService {

    fun calculateDaysFromStart(startDate: LocalDate, referenceDate: LocalDate = LocalDate.now()): Long =
        ChronoUnit.DAYS.between(startDate, referenceDate)

    fun calculateDaysUntil(targetDate: LocalDate, referenceDate: LocalDate = LocalDate.now()): Long =
        ChronoUnit.DAYS.between(referenceDate, targetDate)

    fun getNextAnniversary(startDate: LocalDate, referenceDate: LocalDate = LocalDate.now()): LocalDate {
        var anniversary = startDate.withYear(referenceDate.year)
        if (anniversary.isBefore(referenceDate) || anniversary == referenceDate) {
            anniversary = anniversary.plusYears(1)
        }
        return anniversary
    }

    fun getAnniversaryNumber(startDate: LocalDate, referenceDate: LocalDate = LocalDate.now()): Int =
        referenceDate.year - startDate.year

    fun getUpcomingMilestones(
        startDate: LocalDate,
        referenceDate: LocalDate = LocalDate.now(),
        count: Int = 3
    ): List<Milestone> {
        val daysFromStart = calculateDaysFromStart(startDate, referenceDate)
        val intervals = listOf(100L, 200L, 300L, 365L, 500L, 730L, 1000L, 1095L, 1461L)

        return intervals
            .filter { it > daysFromStart }
            .take(count)
            .map { days ->
                Milestone(
                    days = days,
                    date = startDate.plusDays(days)
                )
            }
    }

    data class Milestone(val days: Long, val date: LocalDate)
}
